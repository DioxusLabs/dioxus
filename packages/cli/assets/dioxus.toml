[application]

# dioxus project name
name = "{{project-name}}"

# default platform
# you can also use `dx serve/build --platform XXX` to use other platform
# value: web | desktop
default_platform = "{{default-platform}}"

# Web `build` & `serve` dist path
out_dir = "dist"

# resource (static) file folder
asset_dir = "public"

# hot reload by default
hot_reload = true

[web.wasm_opt]
# The level wasm-opt should target. z is the smallest. 4 is the fastest.
level = "4"

[web.app]

# HTML title tag content
title = "Dioxus | An elegant GUI library for Rust"

[web.watcher]

index_on_404 = true

watch_path = ["src", "examples"]

[bundler]
# Bundle identifier
identifier = "io.github.{{project-name}}"

# Bundle publisher
publisher = "{{project-name}}"

# Bundle icon
icon = ["icons/icon.png"]

# Bundle resources
resources = ["public/*"]

# Bundle copyright
copyright = ""

# Bundle category
category = "Utility"

# Bundle short description
short_description = "An amazing dioxus application."

# Bundle long description
long_description = """
An amazing dioxus application.
"""
