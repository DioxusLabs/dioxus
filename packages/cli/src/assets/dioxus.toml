[application]

# dioxus project name
name = "{{project-name}}"

# default platfrom
# you can also use `dx serve/build --platform XXX` to use other platform
# value: web | desktop
default_platform = "{{default-platform}}"

# Web `build` & `serve` dist path
out_dir = "dist"

# resource (static) file folder
asset_dir = "public"

[web.app]

# HTML title tag content
title = "Dioxus | An elegant GUI library for Rust"

[web.watcher]

index_on_404 = true

watch_path = ["src"]

# include `assets` in web platform
[web.resource]

# CSS style file
style = []

# Javascript code file
script = []

[web.resource.dev]

# Javascript code file
# serve: [dev-server] only
script = []

[application.plugins]

available = true

required = []
