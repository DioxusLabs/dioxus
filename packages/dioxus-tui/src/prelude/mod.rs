pub use crate::widgets::*;
pub use rink::Config;
