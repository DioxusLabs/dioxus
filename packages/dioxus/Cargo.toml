[package]
name = "dioxus"
version = "0.3.2"
authors = ["Jonathan Kelley", "Dioxus Labs", "ealmloff"]
edition = "2021"
description = "Portable, performant, and ergonomic framework for building cross-platform user interfaces in Rust"
license = "MIT OR Apache-2.0"
repository = "https://github.com/DioxusLabs/dioxus/"
homepage = "https://dioxuslabs.com/docs/0.3/guide/en/index.html"
keywords = ["dom", "ui", "gui", "react", "wasm"]
rust-version = "1.65.0"

[dependencies]
dioxus-core = { workspace = true }
dioxus-html = { workspace = true, optional = true }
dioxus-core-macro = { workspace = true, optional = true }
dioxus-hooks = { workspace = true, optional = true }
dioxus-rsx = { workspace = true, optional = true }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
dioxus-hot-reload = { workspace = true, optional = true }

[features]
default = ["macro", "hooks", "html", "hot-reload"]
macro = ["dioxus-core-macro", "dioxus-rsx"]
html = ["dioxus-html"]
hooks = ["dioxus-hooks"]
hot-reload = ["dioxus-hot-reload"]


[dev-dependencies]
futures-util = { workspace = true }
log = { workspace = true }
rand = { version = "0.8.4", features = ["small_rng"] }
criterion = "0.3.5"
thiserror = "1.0.30"
env_logger = "0.10.0"
tokio = { workspace = true, features = ["full"] }
# dioxus-edit-stream = { workspace = true }


[[bench]]
name = "jsframework"
harness = false
