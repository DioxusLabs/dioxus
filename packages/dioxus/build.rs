fn main() {
    // Warn the user if they enabled the launch feature without any renderers
    if feature_enabled("launch") {
        if feature_enabled("third-party-renderer") {
            return;
        }

        let liveview_renderers = ["liveview", "axum"];
        let fullstack_renderers = ["axum"];
        let client_renderers = ["desktop", "mobile", "web", "tui"];
        let client_renderer_selected = client_renderers
            .iter()
            .any(|renderer| feature_enabled(renderer));
        if feature_enabled("fullstack") {
            let server_fullstack_enabled = fullstack_renderers
                .iter()
                .any(|renderer| feature_enabled(renderer));
            if !server_fullstack_enabled && !client_renderer_selected {
                println!("cargo:warning=You have enabled the launch and fullstack features, but have not enabled any renderers. The application will not be able to launch. Try enabling one of the following renderers: {} for the server or one of the following renderers: {} for the client.", fullstack_renderers.join(", "), client_renderers.join(", "));
            }
        }

        if feature_enabled("liveview") {
            let server_selected = liveview_renderers
                .iter()
                .any(|renderer| feature_enabled(renderer));
            if !server_selected {
                println!("cargo:warning=You have enabled the launch and liveview features, but have not enabled any liveview renderers. The application will not be able to launch. Try enabling one of the following renderers: {}", liveview_renderers.join(", "));
            }
        }

        if !client_renderer_selected {
            println!("cargo:warning=You have enabled the launch feature, but have not enabled any client renderers. The application will not be able to launch. Try enabling one of the following renderers: {}, fullstack or liveview", client_renderers.join(", "));
        }
    }
}

fn feature_enabled(feature: &str) -> bool {
    let feature = "CARGO_FEATURE_".to_owned() + &feature.to_uppercase().replace('-', "_");
    println!("cargo:rerun-if-env-changed={}", feature);
    std::env::var(feature).is_ok()
}
