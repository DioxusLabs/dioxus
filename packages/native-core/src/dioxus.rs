//! Integration between Dioxus and the RealDom

use crate::tree::TreeMut;
use dioxus_core::{BorrowedAttributeValue, ElementId, Mutations, TemplateNode};
use rustc_hash::{FxHashMap, FxHashSet};
use shipyard::Component;

use crate::{
    node::{
        ElementNode, FromAnyValue, NodeType, OwnedAttributeDiscription, OwnedAttributeValue,
        TextNode,
    },
    prelude::*,
    real_dom::NodeTypeMut,
    NodeId,
};

#[derive(Component)]
struct ElementIdComponent(ElementId);

/// The state of the Dioxus integration with the RealDom
pub struct DioxusState {
    templates: FxHashMap<String, Vec<NodeId>>,
    stack: Vec<NodeId>,
    node_id_mapping: Vec<Option<NodeId>>,
}

impl DioxusState {
    /// Initialize the DioxusState in the RealDom
    pub fn create<V: FromAnyValue + Send + Sync>(rdom: &mut RealDom<V>) -> Self {
        let root_id = rdom.root_id();
        let mut root = rdom.get_mut(root_id).unwrap();
        root.insert(ElementIdComponent(ElementId(0)));
        Self {
            templates: FxHashMap::default(),
            stack: vec![root_id],
            node_id_mapping: vec![Some(root_id)],
        }
    }

    /// Convert an ElementId to a NodeId
    pub fn element_to_node_id(&self, element_id: ElementId) -> NodeId {
        self.try_element_to_node_id(element_id).unwrap()
    }

    /// Attempt to convert an ElementId to a NodeId. This will return None if the ElementId is not in the RealDom.
    pub fn try_element_to_node_id(&self, element_id: ElementId) -> Option<NodeId> {
        self.node_id_mapping.get(element_id.0).copied().flatten()
    }

    fn set_element_id<V: FromAnyValue + Send + Sync>(
        &mut self,
        mut node: NodeMut<V>,
        element_id: ElementId,
    ) {
        let node_id = node.id();
        node.insert(ElementIdComponent(element_id));
        if self.node_id_mapping.len() <= element_id.0 {
            self.node_id_mapping.resize(element_id.0 + 1, None);
        }
        self.node_id_mapping[element_id.0] = Some(node_id);
    }

    fn load_child<V: FromAnyValue + Send + Sync>(&self, rdom: &RealDom<V>, path: &[u8]) -> NodeId {
        let mut current = rdom.get(*self.stack.last().unwrap()).unwrap();
        for i in path {
            let new_id = current.child_ids()[*i as usize];
            current = rdom.get(new_id).unwrap();
        }
        current.id()
    }

    /// Updates the dom with some mutations and return a set of nodes that were updated. Pass the dirty nodes to update_state.
    pub fn apply_mutations<V: FromAnyValue + Send + Sync>(
        &mut self,
        rdom: &mut RealDom<V>,
        mutations: Mutations,
    ) {
        for template in mutations.templates {
            let mut template_root_ids = Vec::new();
            for root in template.roots {
                let id = create_template_node(rdom, root);
                template_root_ids.push(id);
            }
            self.templates
                .insert(template.name.to_string(), template_root_ids);
        }

        for e in mutations.edits {
            use dioxus_core::Mutation::*;
            match e {
                AppendChildren { id, m } => {
                    let children = self.stack.split_off(self.stack.len() - m);
                    let parent = self.element_to_node_id(id);
                    for child in children {
                        rdom.get_mut(parent).unwrap().add_child(child);
                    }
                }
                AssignId { path, id } => {
                    let node_id = self.load_child(rdom, path);
                    self.set_element_id(rdom.get_mut(node_id).unwrap(), id);
                }
                CreatePlaceholder { id } => {
                    let node = NodeType::Placeholder;
                    let node = rdom.create_node(node);
                    let node_id = node.id();
                    self.set_element_id(node, id);
                    self.stack.push(node_id);
                }
                CreateTextNode { value, id } => {
                    let node_data = NodeType::Text(TextNode {
                        listeners: FxHashSet::default(),
                        text: value.to_string(),
                    });
                    let node = rdom.create_node(node_data);
                    let node_id = node.id();
                    self.set_element_id(node, id);
                    self.stack.push(node_id);
                }
                HydrateText { path, value, id } => {
                    let node_id = self.load_child(rdom, path);
                    let node = rdom.get_mut(node_id).unwrap();
                    self.set_element_id(node, id);
                    let mut node = rdom.get_mut(node_id).unwrap();
                    let node_type_mut = node.node_type_mut();
                    if let NodeTypeMut::Text(mut text) = node_type_mut {
                        *text.text_mut() = value.to_string();
                    } else {
                        drop(node_type_mut);
                        node.set_type(NodeType::Text(TextNode {
                            text: value.to_string(),
                            listeners: FxHashSet::default(),
                        }));
                    }
                }
                LoadTemplate { name, index, id } => {
                    let template_id = self.templates[name][index];
                    let clone_id = rdom.get_mut(template_id).unwrap().clone_node();
                    let clone = rdom.get_mut(clone_id).unwrap();
                    self.set_element_id(clone, id);
                    self.stack.push(clone_id);
                }
                ReplaceWith { id, m } => {
                    let new_nodes = self.stack.split_off(self.stack.len() - m);
                    let old_node_id = self.element_to_node_id(id);
                    for new in new_nodes {
                        let mut node = rdom.get_mut(new).unwrap();
                        node.insert_before(old_node_id);
                    }
                    rdom.get_mut(old_node_id).unwrap().remove();
                }
                ReplacePlaceholder { path, m } => {
                    let new_nodes = self.stack.split_off(self.stack.len() - m);
                    let old_node_id = self.load_child(rdom, path);
                    for new in new_nodes {
                        let mut node = rdom.get_mut(new).unwrap();
                        node.insert_before(old_node_id);
                    }
                    rdom.get_mut(old_node_id).unwrap().remove();
                }
                InsertAfter { id, m } => {
                    let new_nodes = self.stack.split_off(self.stack.len() - m);
                    let old_node_id = self.element_to_node_id(id);
                    for new in new_nodes.into_iter().rev() {
                        let mut node = rdom.get_mut(new).unwrap();
                        node.insert_after(old_node_id);
                    }
                }
                InsertBefore { id, m } => {
                    let new_nodes = self.stack.split_off(self.stack.len() - m);
                    let old_node_id = self.element_to_node_id(id);
                    for new in new_nodes {
                        rdom.tree_mut().insert_before(old_node_id, new);
                    }
                }
                SetAttribute {
                    name,
                    value,
                    id,
                    ns,
                } => {
                    let node_id = self.element_to_node_id(id);
                    let mut node = rdom.get_mut(node_id).unwrap();
                    let mut node_type_mut = node.node_type_mut();
                    if let NodeTypeMut::Element(element) = &mut node_type_mut {
                        if let BorrowedAttributeValue::None = &value {
                            element.remove_attribute(&OwnedAttributeDiscription {
                                name: name.to_string(),
                                namespace: ns.map(|s| s.to_string()),
                            });
                        } else {
                            element.set_attribute(
                                OwnedAttributeDiscription {
                                    name: name.to_string(),
                                    namespace: ns.map(|s| s.to_string()),
                                },
                                OwnedAttributeValue::from(value),
                            );
                        }
                    }
                }
                SetText { value, id } => {
                    let node_id = self.element_to_node_id(id);
                    let mut node = rdom.get_mut(node_id).unwrap();
                    let node_type_mut = node.node_type_mut();
                    if let NodeTypeMut::Text(mut text) = node_type_mut {
                        *text.text_mut() = value.to_string();
                    }
                }
                NewEventListener { name, id } => {
                    let node_id = self.element_to_node_id(id);
                    let mut node = rdom.get_mut(node_id).unwrap();
                    node.add_event_listener(name);
                }
                RemoveEventListener { id, name } => {
                    let node_id = self.element_to_node_id(id);
                    let mut node = rdom.get_mut(node_id).unwrap();
                    node.remove_event_listener(name);
                }
                Remove { id } => {
                    let node_id = self.element_to_node_id(id);
                    rdom.get_mut(node_id).unwrap().remove();
                }
                PushRoot { id } => {
                    let node_id = self.element_to_node_id(id);
                    self.stack.push(node_id);
                }
            }
        }
    }
}

fn create_template_node<V: FromAnyValue + Send + Sync>(
    rdom: &mut RealDom<V>,
    node: &TemplateNode,
) -> NodeId {
    match node {
        TemplateNode::Element {
            tag,
            namespace,
            attrs,
            children,
        } => {
            let node = NodeType::Element(ElementNode {
                tag: tag.to_string(),
                namespace: namespace.map(|s| s.to_string()),
                attributes: attrs
                    .iter()
                    .filter_map(|attr| match attr {
                        dioxus_core::TemplateAttribute::Static {
                            name,
                            value,
                            namespace,
                        } => Some((
                            OwnedAttributeDiscription {
                                namespace: namespace.map(|s| s.to_string()),
                                name: name.to_string(),
                            },
                            OwnedAttributeValue::Text(value.to_string()),
                        )),
                        dioxus_core::TemplateAttribute::Dynamic { .. } => None,
                    })
                    .collect(),
                listeners: FxHashSet::default(),
            });
            let node_id = rdom.create_node(node).id();
            for child in *children {
                let child_id = create_template_node(rdom, child);
                rdom.get_mut(node_id).unwrap().add_child(child_id);
            }
            node_id
        }
        TemplateNode::Text { text } => rdom
            .create_node(NodeType::Text(TextNode {
                text: text.to_string(),
                ..Default::default()
            }))
            .id(),
        TemplateNode::Dynamic { .. } => rdom.create_node(NodeType::Placeholder).id(),
        TemplateNode::DynamicText { .. } => {
            rdom.create_node(NodeType::Text(TextNode::default())).id()
        }
    }
}

/// A trait that extends the `NodeImmutable` trait with methods that are useful for dioxus.
pub trait NodeImmutableDioxusExt<V: FromAnyValue + Send + Sync>: NodeImmutable<V> {
    /// Returns the id of the element that this node is mounted to.
    /// Not all nodes are mounted to an element, only nodes with dynamic content that have been renderered will have an id.
    fn mounted_id(&self) -> Option<ElementId> {
        let id = self.get::<ElementIdComponent>();
        id.map(|id| id.0)
    }
}

impl<T: NodeImmutable<V>, V: FromAnyValue + Send + Sync> NodeImmutableDioxusExt<V> for T {}
