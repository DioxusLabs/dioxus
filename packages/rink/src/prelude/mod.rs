#[cfg(feature = "dioxus-bindings")]
pub use crate::widgets::*;
