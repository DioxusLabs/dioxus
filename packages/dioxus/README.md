<div style="text-align: center">
  <h1>ðŸŒ—ðŸš€ Dioxus</h1>
  <p>
    <strong>A concurrent, functional, virtual DOM for Rust</strong>
  </p>
</div>

# Resources

This overview provides a brief introduction to Dioxus. For a more in-depth guide, make sure to check out:

- [Getting Started](https://dioxuslabs.com/docs/0.3/guide/en/getting_started/index.html)
- [Book (0.3)](https://dioxuslabs.com/docs/0.3/guide/en/)
- [Examples](https://github.com/DioxusLabs/example-projects)

# Overview and Goals

Dioxus makes it easy to quickly build complex user interfaces with Rust. Any Dioxus app can run in the web browser,
as a desktop app, as a mobile app, or anywhere else provided you build the right renderer.

Dioxus is heavily inspired by React, supporting many of the same concepts:

- Hooks for state
- VirtualDom & diffing
- Concurrency, fibers, and asynchronous rendering
- JSX-like templating syntax

If you know React, then you know Dioxus.

Dioxus is *substantially* more performant than many of the other Rust UI libraries (Yew/Percy) and is *significantly* more performant
than Reactâ€”roughly competitive with InfernoJS.

Remember: Dioxus is a library for declaring interactive user interfacesâ€”it is not a dedicated renderer. Most 1st party renderers for Dioxus currently only support web technologies.

## Brief Overview

All Dioxus apps are built by composing functions that take in a `Scope` which is generic over some `Properties` and return an `Element`.
A `Scope` holds relevant state data for the currently rendered component.

To launch an app, we use the `launch` method for the specific renderer we want to use. In the launch function, we pass the app's `Component`.

```rust, ignore
use dioxus::prelude::*;

fn main() {
    dioxus_desktop::launch(App);
}

// The #[component] attribute streamlines component creation.
// It's not required, but highly recommended. For example, UpperCamelCase components will not generate a warning.
#[component]
fn App(cx: Scope) -> Element {
    cx.render(rsx!("hello world!"))
}
```

## Elements & your first component

To assemble UI trees with Dioxus, you need to use the `render` function on
something called `LazyNodes`. To produce `LazyNodes`, you can use the `rsx!`
macro or the NodeFactory API. For the most part, you want to use the `rsx!`
macro.

Any element in `rsx!` can have attributes, listeners, and children. For
consistency, we force all attributes and listeners to be listed *before*
children.

```rust, ignore
let value = "123";

rsx!(
    div {
        class: "my-class {value}",                  // <--- attribute
        onclick: move |_| log::info!("clicked!"),   // <--- listener
        h1 { "hello world" },                       // <--- child
    }
)
```

The `rsx!` macro accepts attributes in "struct form" and will parse the rest
of the body as child elements and rust expressions. Any rust expression that
implements `IntoIterator<Item = impl IntoVNode>` will be parsed as a child.

```rust, ignore
rsx!(
    div {
        (0..10).map(|_| rsx!(span { "hello world" }))
    }
)
```

Used within components, the `rsx!` macro must be rendered into an `Element` with
the `render` function on Scope.

If we want to omit the boilerplate of `cx.render`, we can simply pass in
`cx` as the first argument of rsx. This is sometimes useful when we need to
render nodes in match statements.

```rust, ignore
#[component[
fn Example(cx: Scope) -> Element {

    // both of these are equivalent
    cx.render(rsx!("hello world"))

    render!("hello world!")
}
```

Putting everything together, we can write a simple component that renders a list of
elements:

```rust, ignore
#[component]
fn App(cx: Scope) -> Element {
    let name = "dave";
    cx.render(rsx!(
        h1 { "Hello, {name}!" }
        div {
            class: "my-class",
            id: "my-id",

            (0..5).map(|i| rsx!(
                div { key: "{i}"
                    "FizzBuzz: {i}"
                }
            ))

        }
    ))
}
```

## Components

We can compose these function components to build a complex app. Each new
component we design must take some Properties. For components with no explicit
properties, we can use the `()` type or simply omit the type altogether.

In Dioxus, all properties are memoized by default!

```rust, ignore
#[component]
fn App(cx: Scope) -> Element {
    cx.render(rsx!(
        Header {
            title: "My App",
            color: "red",
        }
    ))
}
```

Our `Header` component takes a `title` and a `color` property, which we
declare on an explicit `HeaderProps` struct.

```rust, ignore
// The `Props` derive macro lets us add additional functionality to how props are interpreted.
#[derive(Props, PartialEq)]
struct HeaderProps {
    title: String,
    color: String,
}

#[component]
fn Header(cx: Scope<HeaderProps>) -> Element {
    cx.render(rsx!(
        div {
            background_color: "{cx.props.color}"
            h1 { "{cx.props.title}" }
        }
    ))
}
```

The `#[component]` macro also allows you to derive the props
struct from function arguments:

```rust, ignore
#[component]
fn Header(cx: Scope, title: String, color: String) -> Element {
    cx.render(rsx!(
        div {
            background_color: "{color}"
            h1 { "{title}" }
        }
    ))
}
```

Components may also borrow data from their parent component. We just need to
attach some lifetimes to the props struct.
> Note: we don't need to derive `PartialEq` for borrowed props since they cannot be memoized.

```rust, ignore
#[derive(Props)]
struct HeaderProps<'a> {
    title: &'a str,
    color: &'a str,
}

#[component]
fn Header<'a>(cx: Scope<'a, HeaderProps<'a>>) -> Element {
    cx.render(rsx!(
        div {
            background_color: "{cx.props.color}"
            h1 { "{cx.props.title}" }
        }
    ))
}
```

Components that begin with an uppercase letter may be called with
the traditional (for React) curly-brace syntax like so:

```rust, ignore
rsx!(
    Header { title: "My App" }
)
```

Alternatively, if your components begin with a lowercase letter, you can use
the function call syntax:

```rust, ignore
rsx!(
    header( title: "My App" )
)
```

However, the convention is to use UpperCamelCase. The `#[component]` attribute will enforce this,
but you can turn it off if you wish.

## Hooks

While components are reusable forms of UI elements, hooks are reusable forms
of logic. Hooks provide us a way of retrieving state from the `Scope` and using
it to render UI elements.

By convention, all hooks are functions that should start with `use_`. We can
use hooks to define the state and modify it from within listeners.

```rust, ignore
#[component]
fn App(cx: Scope) -> Element {
    let name = use_state(cx, || "world");

    render!("hello {name}!")
}
```

Hooks are sensitive to how they are used. To use hooks, you must abide by the
["rules of hooks" (borrowed from React)](https://reactjs.org/docs/hooks-rules.html):
- Functions with "use_" should not be called in callbacks
- Functions with "use_" should not be called out of order
- Functions with "use_" should not be called in loops or conditionals

In a sense, hooks let us add a field of state to our component without declaring
an explicit state struct. However, this means we need to "load" the struct in the right
order. If that order is wrong, then the hook will pick the wrong state and panic.

Most hooks you'll write are simply compositions of other hooks:

```rust, ignore
fn use_username(cx: &ScopeState, id: Uuid) -> bool {
    let users = use_context::<Users>(cx);
    users.get(&id).map(|user| user.logged_in).ok_or(false)
}
```

To create entirely new foundational hooks, we can use the `use_hook` method on `ScopeState`.

```rust, ignore
fn use_mut_string(cx: &ScopeState) -> &mut String {
    cx.use_hook(|_| "Hello".to_string())
}
```

If you want to extend Dioxus with some new functionality, you'll probably want to implement a new hook from scratch.

## Putting it all together

Using components, templates, and hooks, we can build a simple app.

```rust, ignore
use dioxus::prelude::*;

fn main() {
    dioxus_desktop::launch(App);
}

#[component]
fn App(cx: Scope) -> Element {
    let count = use_state(cx, || 0);

    cx.render(rsx!(
        div { "Count: {count}" }
        button { onclick: move |_| count.set(count + 1), "Increment" }
        button { onclick: move |_| count.set(count - 1), "Decrement" }
    ))
}
```

## Features

This overview doesn't cover everything. Make sure to check out the tutorial and reference guide on the official
website for more details.

Beyond this overview, Dioxus supports:
- Server-side rendering
- Concurrent rendering (with async support)
- Web/Desktop/Mobile support
- Pre-rendering and rehydration
- Fragments, Portals, and Suspense
- Inline-styles
- Custom event handlers
- Custom elements
- Basic fine-grained reactivity (IE SolidJS/Svelte)
- and more!

Good luck!

## Inspiration, Resources, Alternatives, and Credits

Dioxus is inspired by:
- React: for its hooks, concurrency, suspense
- Dodrio: for its research in bump allocation, double buffering, and diffing architecture

Alternatives to Dioxus include:
- Yew: supports function components and web, but no SSR, borrowed data, or bump allocation. Rather slow at times.
- Percy: supports function components, web, ssr, but lacks state management
- Sycamore: supports function components, web, ssr, but is closer to SolidJS than React
- MoonZoom/Seed: opinionated frameworks based on the Elm model (message, update)â€”no hooks

We've put a lot of work into making Dioxus ergonomic and *familiar*.
Our target audience is TypeScript developers looking to switch to Rust for the webâ€”so we need to be comparable to React.
