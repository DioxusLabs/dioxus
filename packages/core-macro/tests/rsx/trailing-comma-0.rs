// Given an `rsx!` invocation with a missing trailing comma,
// ensure the stderr output has an informative span.

use dioxus::prelude::*;

fn main() {
    rsx! {
        p {
            class: "foo bar"
            "Hello world"
        }
    };
}
