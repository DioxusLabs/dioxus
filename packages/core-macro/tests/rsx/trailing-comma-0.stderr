error: missing trailing comma
 --> tests/rsx/trailing-comma-0.rs:9:20
  |
9 |             class: "foo bar"
  |                    ^^^^^^^^^
